lib/msal-common/src/client/DeviceCodeClient.ts

```diff
@@ -137,72 +137,73 @@ export class DeviceCodeClient extends BaseClient {
-
-                        this.logger.error(`User defined timeout for device code polling reached. The timeout was set for ${userSpecifiedTimeout}`);
-                        clearInterval(intervalId);
-                        reject(ClientAuthError.createUserTimeoutReachedError());
-
-                    } else if (TimeUtils.nowSeconds() > deviceCodeExpirationTime) {
-
-                        if (userSpecifiedTimeout) {
-                            this.logger.verbose(`User specified timeout ignored as the device code has expired before the timeout elapsed. The user specified timeout was set for ${userSpecifiedTimeout}`);
-                        }
-
-                        this.logger.error(`Device code expired. Expiration time of device code was ${deviceCodeExpirationTime}`);
-                        clearInterval(intervalId);
-                        reject(ClientAuthError.createDeviceCodeExpiredError());
-
-                    } else {
-                        const thumbprint: RequestThumbprint = {
-                            clientId: this.config.authOptions.clientId,
-                            authority: request.authority,
-                            scopes: request.scopes
-                        };
-                        const response = await this.executePostToTokenEndpoint(
-                            this.authority.tokenEndpoint,
-                            requestBody,
-                            headers,
-                            thumbprint);
-
-                        if (response.body && response.body.error === Constants.AUTHORIZATION_PENDING) {
-                            // user authorization is pending. Sleep for polling interval and try again
-                            this.logger.info(response.body.error_description || "no_error_description");
-                        } else {
-                            clearInterval(intervalId);
-                            resolve(response.body);
-                        }
-                    }
-                } catch (error) {
-                    clearInterval(intervalId);
-                    reject(error);
-                }
-            }, pollingIntervalMilli);
-        });
+      request: CommonDeviceCodeRequest,
+      deviceCodeResponse: DeviceCodeResponse): Promise<ServerAuthorizationTokenResponse> {
+
+      const requestBody = this.createTokenRequestBody(request, deviceCodeResponse);
+      const headers: Record<string, string> = this.createDefaultTokenRequestHeaders();
+
+      const userSpecifiedTimeout = request.timeout ? TimeUtils.nowSeconds() + request.timeout : undefined;
+      const deviceCodeExpirationTime = TimeUtils.nowSeconds() + deviceCodeResponse.expiresIn;
+      const pollingIntervalMilli = deviceCodeResponse.interval * 1000;
+
+      /**
+       * Breaks the polling with specific conditions.
+       */
+      const continuePolling = (): boolean => {
+        if (request.cancel) {
```

Comment:
FYI our lint rules ask for 4 spaces, instead of 2, which is why this diff looks so big.

Reply:
Fixed this and now the diff is as expected.

Reply:
Thanks!